README.md
## python_ethernet_rmp (indigo) - 0.0.2-0

The packages in the `python_ethernet_rmp` repository were released into the `indigo` distro by running `/usr/bin/bloom-release --rosdistro indigo --track indigo python_ethernet_rmp` on `Tue, 29 Jul 2014 16:43:40 -0000`

The `python_ethernet_rmp` package was released.

Version of package(s) in repository `python_ethernet_rmp`:
- rosdistro version: `0.0.1-0`
- old version: `0.0.1-0`
- new version: `0.0.2-0`

Versions of tools used:
- bloom version: `0.5.10`
- catkin_pkg version: `0.2.4`
- rosdep version: `0.10.28`
- rosdistro version: `0.3.5`
- vcstools version: `0.1.35`


## python_ethernet_rmp (indigo) - 0.0.1-0

The packages in the `python_ethernet_rmp` repository were released into the `indigo` distro by running `/usr/bin/bloom-release --rosdistro indigo --track indigo python_ethernet_rmp --new-track` on `Tue, 10 Jun 2014 12:50:36 -0000`

The `python_ethernet_rmp` package was released.

Version of package(s) in repository `python_ethernet_rmp`:
- rosdistro version: `null`
- old version: `null`
- new version: `0.0.1-0`

Versions of tools used:
- bloom version: `0.5.9`
- catkin_pkg version: `0.2.2`
- rosdep version: `0.10.27`
- rosdistro version: `0.3.5`
- vcstools version: `0.1.35`


python_ethernet_rmp-release
===========================


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package python_ethernet_rmp
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.2 (2014-07-29)
------------------
* Merge pull request `#6 <https://github.com/WPI-RAIL/python_ethernet_rmp/issues/6>`_ from cmdunkers/develop
  fixed the default_bitmap issue
* Fixed the default_config bitmap to 1 instead of the default since default does not work with carl
* Merge pull request `#2 <https://github.com/WPI-RAIL/python_ethernet_rmp/issues/2>`_ from WPI-RAIL/develop
  Update CMakeLists.txt
* Merge pull request `#1 <https://github.com/WPI-RAIL/python_ethernet_rmp/issues/1>`_ from WPI-RAIL/master
  Cleanup and release of 0.0.1
* Update CMakeLists.txt
* Merge pull request `#5 <https://github.com/WPI-RAIL/python_ethernet_rmp/issues/5>`_ from WPI-RAIL/develop
  0.0.1
* Contributors: Chris Dunkers, Russell Toris, cmdunkers

0.0.1 (2014-06-10)
------------------
* cleanup
* travis and cleanup
* Merge pull request #4 from cmdunkers/develop
  Fixed Authors
* Fixed Authros
* Merge pull request #3 from cmdunkers/develop
  basic cleanup completed
* cleanup
* basic cleanup
* Changed minor things
* Merge pull request #1 from cmdunkers/develop
  Added and modified the Segway communication code to fit our needs
* worked on modifying rmp_config_params so the dynamic varibales are the ones read in
* Made it so there is only feedback received when TX is called
* added and modified the provided segway example
* Initial commit
* Contributors: Chris Dunkers, Russell Toris


src/python_ethernet_rmp/crc16.py
"""--------------------------------------------------------------------
COPYRIGHT 2013 SEGWAY Inc.

Software License Agreement:

The software supplied herewith by Segway Inc. (the "Company") for its 
RMP Robotic Platforms is intended and supplied to you, the Company's 
customer, for use solely and exclusively with Segway products. The 
software is owned by the Company and/or its supplier, and is protected 
under applicable copyright laws.  All rights are reserved. Any use in 
violation of the foregoing restrictions may subject the user to criminal 
sanctions under applicable laws, as well as to civil liability for the 
breach of the terms and conditions of this license. The Company may 
immediately terminate this Agreement upon your use of the software with 
any products that are not Segway products.

The software was written using Python programming language.  Your use 
of the software is therefore subject to the terms and conditions of the 
OSI- approved open source license viewable at http://www.python.org/.  
You are solely responsible for ensuring your compliance with the Python 
open source license.

You shall indemnify, defend and hold the Company harmless from any claims, 
demands,liabilities or expenses, including reasonable attorneys fees, incurred 
by the Company as a result of any claim or proceeding against the Company 
arising out of or based upon: 

(i) The combination, operation or use of the software by you with any hardware, 
    products, programs or data not supplied or approved in writing by the Company, 
    if such claim or proceeding would have been avoided but for such combination, 
    operation or use.
 
(ii) The modification of the software by or on behalf of you 

(iii) Your use of the software.

 THIS SOFTWARE IS PROVIDED IN AN "AS IS" CONDITION. NO WARRANTIES,
 WHETHER EXPRESS, IMPLIED OR STATUTORY, INCLUDING, BUT NOT LIMITED
 TO, IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 PARTICULAR PURPOSE APPLY TO THIS SOFTWARE. THE COMPANY SHALL NOT,
 IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL OR
 CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER.
 
 \file   crc16.py

 \brief  This module contains a collection of functions for calculating
         a CRC-16.

 \Platform: Cross Platform
--------------------------------------------------------------------"""

"""
CRC16 defines
"""
CRC_ADJUSTMENT = 0xA001;
CRC_TABLE_SIZE = 256;
INITIAL_CRC = 0;

HIGH_BYTE_MASK  = 0xFF00;
LOW_BYTE_MASK   = 0x00FF;
MOVE_BYTE_SHIFT = 8;
LS_BIT          = 0x0001;
BITS_PER_BYTE   = 8;

crc_table= [0]*CRC_TABLE_SIZE;

"""
This creates the CRC table
"""
def generate_crc_table():

    for x in range(0,CRC_TABLE_SIZE):

        table_value = 0;
        k = x;

        for j in range(0,BITS_PER_BYTE):
            if (((table_value ^ k) & LS_BIT) == LS_BIT):
                table_value = (table_value >> 1) ^ CRC_ADJUSTMENT;
            else:
                table_value >>= 1;
            k >>= 1;

        crc_table[x] = table_value;

"""
This computes an updated CRC 16 given the current value of the CRC 16 and
a new data byte.
"""
def calculate_crc_16(old_crc,new_word):

    temp = old_crc ^ new_word;
    new_crc = (old_crc >> MOVE_BYTE_SHIFT) ^ crc_table[temp & LOW_BYTE_MASK];

    return new_crc;

"""
This computes the CRC for a buffer passed in
"""
def compute_buffer_crc(byte_buffer, bytes_in_buffer):
    crc_index = bytes_in_buffer - 2;
    new_crc = 0;

    """
    We'll loop through each word of the message and update
    the CRC.  Start with the value chosen for CRC initialization.
    """
    for x in range(0,crc_index):
        """
        Now we'll send each byte to the CRC calculation.
        """
        new_crc = calculate_crc_16(new_crc, byte_buffer[x]);


    byte_buffer[crc_index]   = int((new_crc & HIGH_BYTE_MASK) >> MOVE_BYTE_SHIFT);
    byte_buffer[crc_index+1] = int((new_crc & LOW_BYTE_MASK));
    
"""
This computes the CRC for a buffer passed in and checks it against
the one contained in the buffer
"""
def buffer_crc_is_valid(byte_buffer, bytes_in_buffer):
    
    crc_index = bytes_in_buffer - 2;
    new_crc = 0;
    success = False;

    """
    We'll loop through each word of the message and update
    the CRC.  Start with the value chosen for CRC initialization.
    """
    for x in range(0,crc_index):
        new_crc = calculate_crc_16(new_crc, byte_buffer[x]);

    """
    The new CRC is checked against that stored in the buffer.
    """
    received_crc = ((byte_buffer[crc_index] << MOVE_BYTE_SHIFT) & HIGH_BYTE_MASK);
    received_crc |= (byte_buffer[crc_index+1] & LOW_BYTE_MASK);
    
    if (received_crc == new_crc):
        success = True;
    else:
        success = False;

    return (success);


src/python_ethernet_rmp/utils.py
"""--------------------------------------------------------------------
COPYRIGHT 2013 SEGWAY Inc.

Software License Agreement:

The software supplied herewith by Segway Inc. (the "Company") for its 
RMP Robotic Platforms is intended and supplied to you, the Company's 
customer, for use solely and exclusively with Segway products. The 
software is owned by the Company and/or its supplier, and is protected 
under applicable copyright laws.  All rights are reserved. Any use in 
violation of the foregoing restrictions may subject the user to criminal 
sanctions under applicable laws, as well as to civil liability for the 
breach of the terms and conditions of this license. The Company may 
immediately terminate this Agreement upon your use of the software with 
any products that are not Segway products.

The software was written using Python programming language.  Your use 
of the software is therefore subject to the terms and conditions of the 
OSI- approved open source license viewable at http://www.python.org/.  
You are solely responsible for ensuring your compliance with the Python 
open source license.

You shall indemnify, defend and hold the Company harmless from any claims, 
demands, liabilities or expenses, including reasonable attorneys fees, incurred 
by the Company as a result of any claim or proceeding against the Company 
arising out of or based upon: 

(i) The combination, operation or use of the software by you with any hardware, 
    products, programs or data not supplied or approved in writing by the Company, 
    if such claim or proceeding would have been avoided but for such combination, 
    operation or use.
 
(ii) The modification of the software by or on behalf of you 

(iii) Your use of the software.

 THIS SOFTWARE IS PROVIDED IN AN "AS IS" CONDITION. NO WARRANTIES,
 WHETHER EXPRESS, IMPLIED OR STATUTORY, INCLUDING, BUT NOT LIMITED
 TO, IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 PARTICULAR PURPOSE APPLY TO THIS SOFTWARE. THE COMPANY SHALL NOT,
 IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL OR
 CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER.
 
 \file   utils.py

 \brief  This module contains general utility functions

 \Platform: Cross Platform
--------------------------------------------------------------------"""
import struct
import socket

"""
For IEEE754 processors this function converts a 32-bit floating point number to
a 32-bit integer representation
"""
def convert_float_to_u32(value):
    return struct.unpack('I', struct.pack('f', value))[0]

"""
For IEEE754 processors this function converts a 32-bit integer representation
of a floating point value to float representation
"""
def convert_u32_to_float(bits):
    return struct.unpack('f', struct.pack('I', bits))[0]

"""
Used to convert a byte array (string) into an array of 32-bit values
"""
def convert_byte_data_to_U32(data):

    rx_dat = [];
    k = 0;
    
    #
    # Convert the string into a byte array
    #
    
    for x in range(0,len(data)):
        rx_dat.append(ord(data[x]));
        
    number_of_u32s = (len(rx_dat)/4)

    #
    # Convert the byte array into an array of 32bit values
    #
    converted = [0]*number_of_u32s;
    for x in range(0,number_of_u32s):
        converted[x] = int((((rx_dat[k]   << 24) & 0xFF000000)) |
                        (((rx_dat[k+1] << 16) & 0x00FF0000)) |
                        (((rx_dat[k+2] << 8)  & 0x0000FF00)) |
                          (rx_dat[k+3] & 0x000000FF));

        k+=4;
        
    return converted;

"""
Used to convert an IP address string in dotted quad format to an integer
"""  
def dottedQuadToNum(ip):
    "convert decimal dotted quad string to long integer"
    return struct.unpack('I',socket.inet_aton(ip))[0]

"""
Used to convert an IP address in integer format to a dotted quad string
""" 
def numToDottedQuad(n):
    "convert long int to dotted quad string"
    return socket.inet_ntoa(struct.pack('I',n))


src/python_ethernet_rmp/io_eth_cmd.py
"""--------------------------------------------------------------------
COPYRIGHT 2013 SEGWAY Inc.

Software License Agreement:

The software supplied herewith by Segway Inc. (the "Company") for its 
RMP Robotic Platforms is intended and supplied to you, the Company's 
customer, for use solely and exclusively with Segway products. The 
software is owned by the Company and/or its supplier, and is protected 
under applicable copyright laws.  All rights are reserved. Any use in 
violation of the foregoing restrictions may subject the user to criminal 
sanctions under applicable laws, as well as to civil liability for the 
breach of the terms and conditions of this license. The Company may 
immediately terminate this Agreement upon your use of the software with 
any products that are not Segway products.

The software was written using Python programming language.  Your use 
of the software is therefore subject to the terms and conditions of the 
OSI- approved open source license viewable at http://www.python.org/.  
You are solely responsible for ensuring your compliance with the Python 
open source license.

You shall indemnify, defend and hold the Company harmless from any claims, 
demands,liabilities or expenses, including reasonable attorneys fees, incurred 
by the Company as a result of any claim or proceeding against the Company 
arising out of or based upon: 

(i) The combination, operation or use of the software by you with any hardware, 
    products, programs or data not supplied or approved in writing by the Company, 
    if such claim or proceeding would have been avoided but for such combination, 
    operation or use.
 
(ii) The modification of the software by or on behalf of you 

(iii) Your use of the software.

 THIS SOFTWARE IS PROVIDED IN AN "AS IS" CONDITION. NO WARRANTIES,
 WHETHER EXPRESS, IMPLIED OR STATUTORY, INCLUDING, BUT NOT LIMITED
 TO, IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 PARTICULAR PURPOSE APPLY TO THIS SOFTWARE. THE COMPANY SHALL NOT,
 IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL OR
 CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER.
 
 \file   io_eth_cmd.py

 \brief  This module contains the ethernet UDP communication protocol

 \Platform: Cross platform
--------------------------------------------------------------------"""
from utils import convert_byte_data_to_U32
import socket

class IO_ETHERNET:
    def __init__(self,rmp_address):
        
        self.success = True
        
        """
        Initialize link parameters
        """
        my_address = ('',rmp_address[1])

        """
        Initialize the UDP connection
        """
        self.conn = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.conn.setblocking(False)
        self.conn.bind(my_address)
        try:
            self.conn.connect(rmp_address)
        except:
            self.success = False

    def Send(self,data):
        try:
            self.conn.sendall(data)            
        except:
            pass
    
    def Receive(self,num_of_return):
        
        """
        The number of bytes expected is the number of 32-bit messages times
        the bytes per 32-bit word
        """    
        num_of_bytes = num_of_return * 4
        
        """
        Try receiving the data up to a maximum size. If it fails
        empty the data
        """
        try:
            data = self.conn.recv(1024)
        except:
            data = []
            
        """
        If the data is the length expected, convert the byte data to U32 data and return it.
        Otherwise return the None type.
        """
        if (len(data) == num_of_bytes):
            return_data = convert_byte_data_to_U32(data);
        else:
            return_data = None; 
            
        return return_data;

    def Close(self):
        self.conn.shutdown(socket.SHUT_RDWR)
        self.conn.close()


src/python_ethernet_rmp/rmp_interface.py
"""--------------------------------------------------------------------
COPYRIGHT 2013 SEGWAY Inc.

Software License Agreement:

The software supplied herewith by Segway Inc. (the "Company") for its 
RMP Robotic Platforms is intended and supplied to you, the Company's 
customer, for use solely and exclusively with Segway products. The 
software is owned by the Company and/or its supplier, and is protected 
under applicable copyright laws.  All rights are reserved. Any use in 
violation of the foregoing restrictions may subject the user to criminal 
sanctions under applicable laws, as well as to civil liability for the 
breach of the terms and conditions of this license. The Company may 
immediately terminate this Agreement upon your use of the software with 
any products that are not Segway products.

The software was written using Python programming language.  Your use 
of the software is therefore subject to the terms and conditions of the 
OSI- approved open source license viewable at http://www.python.org/.  
You are solely responsible for ensuring your compliance with the Python 
open source license.

You shall indemnify, defend and hold the Company harmless from any claims, 
demands, liabilities or expenses, including reasonable attorneys fees, incurred 
by the Company as a result of any claim or proceeding against the Company 
arising out of or based upon: 

(i) The combination, operation or use of the software by you with any hardware, 
    products, programs or data not supplied or approved in writing by the Company, 
    if such claim or proceeding would have been avoided but for such combination, 
    operation or use.
 
(ii) The modification of the software by or on behalf of you 

(iii) Your use of the software.

 THIS SOFTWARE IS PROVIDED IN AN "AS IS" CONDITION. NO WARRANTIES,
 WHETHER EXPRESS, IMPLIED OR STATUTORY, INCLUDING, BUT NOT LIMITED
 TO, IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 PARTICULAR PURPOSE APPLY TO THIS SOFTWARE. THE COMPANY SHALL NOT,
 IN ANY CIRCUMSTANCES, BE LIABLE FOR SPECIAL, INCIDENTAL OR
 CONSEQUENTIAL DAMAGES, FOR ANY REASON WHATSOEVER.
 
 \file   rmp_interface.py

 \brief  This module contains the interface to the RMP

 \Platform: Cross Platform
--------------------------------------------------------------------"""
from crc16  import *
from utils import *
from system_defines import *
from rmp_config_params import *
from user_event_handlers import *
from io_eth_cmd import IO_ETHERNET
import time,os,sys


"""
Define the variable for fixed point Q15 format for
converting certain parameters for OMNI
"""
Q15             = 32767

"""
Main class for the RMP interface
"""
class RMP:
    def __init__(self,rmp_addr,rsp_queue,cmd_queue,in_flags,out_flags,update_rate=MIN_UPDATE_PERIOD_SEC,log_data=False):        
        """
        generate the CRC table
        """
        generate_crc_table()        
        
        """
        Create a dictionary for the feedback
        """
        self.user_defined_feedback = dict()
        
        """
        Indicates a logfile has been created
        """
        self.logfile_started = False
        
        """
        Initialize the bitmaps and create a dictionary for holding the
        user defined feedback data
        """
        self.bitmap = [0]*4
        self.bitmap[0] = CONFIG_PARAMS[RMP_CMD_SET_USER_FB_1_BITMAP-1][2]
        self.bitmap[1] = CONFIG_PARAMS[RMP_CMD_SET_USER_FB_2_BITMAP-1][2]
        self.bitmap[2] = CONFIG_PARAMS[RMP_CMD_SET_USER_FB_3_BITMAP-1][2]
        self.bitmap[3] = CONFIG_PARAMS[RMP_CMD_SET_USER_FB_4_BITMAP-1][2]
        
        """
        Create arrays of bitmaps defining the type of feedback contained
        in each UDFB
        """
        self.fp_mask  = [RMP_FLOATING_POINT_FEEDBACK_1_MASK,
                         RMP_FLOATING_POINT_FEEDBACK_2_MASK,
                         RMP_FLOATING_POINT_FEEDBACK_3_MASK,
                         RMP_FLOATING_POINT_FEEDBACK_4_MASK]
        
        self.hex_mask  = [RMP_HEX_FEEDBACK_1_MASK,
                          RMP_HEX_FEEDBACK_2_MASK,
                          RMP_HEX_FEEDBACK_3_MASK,
                          RMP_HEX_FEEDBACK_4_MASK]
        
        self.ip_mask  = [RMP_IP_FEEDBACK_1_MASK,
                         RMP_IP_FEEDBACK_2_MASK,
                         RMP_IP_FEEDBACK_3_MASK,
                         RMP_IP_FEEDBACK_4_MASK]
        
        """
        Create a header dictionary array
        """
        self.hdr_dicts = [feedback_1_bitmap_menu_items,
                          feedback_2_bitmap_menu_items,
                          feedback_3_bitmap_menu_items,
                          feedback_4_bitmap_menu_items]
        
        """
        Update the local feedback dictionary
        """
        self.update_feedback_dict(None,True)
        
        """
        Get the RMP address
        """
        self.comm = IO_ETHERNET(rmp_addr)
        
        self.in_flags = in_flags
        self.out_flags = out_flags
        
        if (update_rate >= MIN_UPDATE_PERIOD_SEC):
            self.delay = update_rate
        else:
            print "Bad Update Period needs to be longer than 0.01s....."
            print "exiting......"
            self.out_flags.put(RMP_INIT_FAILED)
            self.Close()
                    
        if (False == self.comm.success):
            print "Could not connect to RMP UDP socket....."
            print "exiting......"
            self.out_flags.put(RMP_INIT_FAILED)
            self.Close()
            
        if (False == self.set_and_verify_config_params(CONFIG_PARAMS)):
            print "Could not configure RMP......"
            print "exiting......"
            self.out_flags.put(RMP_INIT_FAILED)
            self.Close()
        
        """
        Get the queues and the time stamps now that initialization was successful
        """
        self.cmd_queue = cmd_queue
        self.rsp_queue = rsp_queue
        self.last_update_time = time.time()
        self.log_file_start_time = time.time()
        
        """
        If the caller wants to log data check for the log directory, create the file and write
        the header.
        """
        self.log_data = log_data
        if self.log_data:
            dirpath = os.getcwd()+"/RMP_DATA_LOGS"
            if (False == os.path.exists(dirpath)):
                os.mkdir(dirpath)
            filename = dirpath + "/" + "RMP_DATA_LOG_" + time.strftime("%m%d%Y_%H%M%S") + ".csv"
            self.logfile = open(filename,'w')
            self.logfile.write(','.join(self.log_file_header))
            self.logfile_started = True 
            
        """
        Run the thread
        """
        self.run()
        
    def run(self):
        while True:
            
            """
            Check the flags coming in. Presently there is only a kill.
            """
            while not self.in_flags.empty():
                if (RMP_KILL == self.in_flags.get()):
                    print "RMP thread has been killed by application......"
                    print "exiting........."
                    self.out_flags.put(RMP_IS_DEAD)
                    self.Close()
                    sys.exit() 

            """
            If enough time has passed since the last transmit, check the command
            queue and send the next command. Then set the TX_RDY flag so the parent
            knows that the queue can be repopulated
            """
            if ((time.time() - self.last_update_time) > self.delay):
                if not self.cmd_queue.empty():
                    self.update_rmp_commands(self.cmd_queue.get())
                    
                self.out_flags.put(RMP_TX_RDY)
                self.last_update_time = time.time()

            """
            Check for data each time and put it in the queue is it exists.
            """
            data = self.comm.Receive(self.expected_items)
            if (self.update_feedback_dict(data,False)):
                self.rsp_queue.put(self.user_defined_feedback)
                self.out_flags.put(RMP_RSP_DATA_RDY)

    def set_and_verify_config_params(self,config):
        """
        The commands that force the feedback array to just contain the configurable elements
        but not update the UDFB to do so. This allows the user to verify the configuration
        while it is changing.
        """
        force_nvm_feedback = [RMP_CFG_CMD_ID,RMP_CMD_FORCE_CONFIG_FEEDBACK_BITMAPS,1]
        set_user_feedback = [RMP_CFG_CMD_ID,RMP_CMD_FORCE_CONFIG_FEEDBACK_BITMAPS,0]
                
        """
        Start by sending the force config feedback command and check for the
        appropriate response
        """
        attempts = 0
        success = False
        while ((False == success) and (attempts<10)):
            self.update_rmp_commands(force_nvm_feedback)
            time.sleep(0.05)
            loaded_params = self.comm.Receive(FORCED_CONFIG_FEEDBACK_ITEMS)
            if (None != loaded_params):
                success = True
            else:
                attempts += 1
        
        """
        If the command to force the config feedback was successful check to see if there
        are non-matching configuration parameters
        """
        if (False == success):
            print "Could not set RMP_CMD_FORCE_CONFIG_FEEDBACK_BITMAPS....."
            print "The platform did not respond, ensure it is operational and the IP address is correct...."
            return False
        else:
            non_matching_params = []
            for i in range(0,NUMBER_OF_NVM_CONFIG_PARAMS):
                if (loaded_params[i] != config[i][2]):
                    non_matching_params.append(i)
        
        """
        Load each configuration parameter which does not match the configuration file
        """
        attempts = 0    
        for i in range(0,len(non_matching_params)):
            idx = non_matching_params[i]
            success = False
            while ((False == success) and (attempts<10)):
                attempts+=1
                self.update_rmp_commands(config[idx])
                time.sleep(0.05)
                loaded_params = self.comm.Receive(FORCED_CONFIG_FEEDBACK_ITEMS)
                if (None != loaded_params):
                    if (loaded_params[idx] == config[idx][2]):
                        success = True
                        attempts = 0
            if (False == success):
                break
        
        """
        Notify the user if we could not set the parameter
        """
        if (False == success):
            print "Could not set param %(1)s....." %{"1":config_param_dict[idx+1]}
            print "The parameter is likely not valid, check it in rmp_config_params.py"
            return False
        
        """
        Switch back to the user feedback array and make sure we get an appropriate response
        """
        attempts = 0
        success = False
        while ((False == success) and (attempts<10)):
            self.update_rmp_commands(set_user_feedback)
            time.sleep(0.05)
            data = self.comm.Receive(self.expected_items)
            if (None != data):
                success = True
            else:
                attempts += 1
                    
        """
        Could not reset the feedback 
        """
        if (False == success):
            print "Could not set user defined feedback" %{"1":idx+1}
            print "The platform did not respond, "
            return False
        
        return True
    
    def update_rmp_commands(self,input_cmd):
       
        """
        Populate the message to the RMP platform if it is not a
        valid format return False
        """
        try:
            cmds = [0]*3
            cmds[0] = input_cmd[0]
            send_cmd = True
            
            if (cmds[0] == RMP_OMNI_MOTION_CMD_ID):
                vel_cmd = int((input_cmd[1] * Q15))
                yaw_cmd = int((input_cmd[2] * Q15))
                cmds[1]= (((vel_cmd << 16) & 0xFFFF0000) | (yaw_cmd & 0x0000FFFF))
                cmds[2] = int(convert_float_to_u32(input_cmd[3]))
            elif (cmds[0] == RMP_MOTION_CMD_ID):
                cmds[1] = int(convert_float_to_u32(input_cmd[1]))
                cmds[2] = int(convert_float_to_u32(input_cmd[2]))
            elif (cmds[0] == RMP_CFG_CMD_ID):
                cmds[1] = int(input_cmd[1])
                cmds[2] = int(input_cmd[2])
            else:
                send_cmd = False
        except:
            send_cmd = False
            
        """
        If we have a valid command send it 
        """
        if (True == send_cmd):
                
            output = self.Convert_RMP_Cmds_for_Serial_Interface(cmds)
            self.comm.Send(output)
            
        return send_cmd
    
    def update_feedback_dict(self,data=None,init=False):
        ret = True
        
        """
        Determine the way the variables should be converted to human readable form. There are
        four ways, floating point, hex, IP strings, and the default is integer
        """
        item = 0
        if init:
            self.log_file_header = ["Time_stamp"]
            for x in range(0,4):                
                for i in range(0,32):
                    if (self.bitmap[x] & (1<<i)):
                        if (self.fp_mask[x] & (1<<i)):
                            self.log_file_header.append(self.hdr_dicts[x][(1<<i)])
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = 0.0
                        elif (self.hex_mask[x] & (1 << i)):
                            self.log_file_header.append(self.hdr_dicts[x][(1<<i)])
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = hex(0)
                        elif (self.ip_mask[x] & (1 << i)):
                            self.log_file_header.append(self.hdr_dicts[x][(1<<i)])
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = numToDottedQuad(0)
                        else:
                            self.log_file_header.append(self.hdr_dicts[x][(1<<i)])
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = 0
                        
                        item +=1
            """
            Add one for the CRC
            """
            self.expected_items = item+1
            
            """
            Append a line end to the logfile header
            """
            self.log_file_header.append("\n")
            
        elif (data != None):
            """
            If we have data convert it and add it to the dictionary
            """
            temp = []
            temp.append(str(time.time() - self.log_file_start_time))
            self.user_defined_feedback.clear();
            for x in range(0,4):
                for i in range(0,32):
                    if (self.bitmap[x] & (1<<i)): 
                        if (self.fp_mask[x] & (1<<i)):
                            temp.append(str(round(convert_u32_to_float(data[item]),3)))
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = round(convert_u32_to_float(data[item]),3);
                        elif (self.hex_mask[x] & (1 << i)):
                            temp.append("0x%(1)08X" %{"1":data[item]})
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = hex(data[item]);
                        elif (self.ip_mask[x] & (1 << i)):
                            temp.append(str(numToDottedQuad(data[item])))
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = numToDottedQuad(data[item]);
                        else:
                            temp.append(str(int(data[item])))
                            self.user_defined_feedback[self.hdr_dicts[x][(1<<i)]] = int(data[item]);
                        
                        item += 1
            
            """
            Log the data if called for
            """
            if(True == self.logfile_started):
                temp.append("\n")
                self.logfile.write(",".join(temp))
        else:
            ret = False
        
        return ret
            
    def Close(self):
        if (True == self.logfile_started):
            self.logfile.close()
        self.comm.Close()
        sys.exit()
            
    def Convert_RMP_Cmds_for_Serial_Interface(self,cmds):
        """
        Convert a set of commands for the UDP Ethernet interface
        """
        rmp_cmd = [0]*NUM_USB_ETH_BYTES;
        
        rmp_cmd[RMP_USB_ETH_CAN_ID_HIGH_INDEX] = int((cmds[0] & 0xFF00) >> 8)
        rmp_cmd[RMP_USB_ETH_CAN_ID_LOW_INDEX]  = int((cmds[0] & 0x00FF))
        rmp_cmd[RMP_USB_ETH_CAN_DATA_0_INDEX]  = int((cmds[1] & 0xFF000000) >> 24)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_1_INDEX]  = int((cmds[1] & 0x00FF0000) >> 16)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_2_INDEX]  = int((cmds[1] & 0x0000FF00) >> 8)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_3_INDEX]  = int((cmds[1] & 0x000000FF))
        rmp_cmd[RMP_USB_ETH_CAN_DATA_4_INDEX]  = int((cmds[2] & 0xFF000000) >> 24)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_5_INDEX]  = int((cmds[2] & 0x00FF0000) >> 16)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_6_INDEX]  = int((cmds[2] & 0x0000FF00) >> 8)
        rmp_cmd[RMP_USB_ETH_CAN_DATA_7_INDEX]  = int((cmds[2] & 0x000000FF))
        
        """
        Compute the CRC for the command 
        """
        compute_buffer_crc(rmp_cmd,NUM_USB_ETH_BYTES)
        
        """
        Convert the string to char data and return it
        """
        rmp_cmd_chars = []
        for x in range(0,len(rmp_cmd)):
            rmp_cmd_chars.append(chr(rmp_cmd[x]))   
        
        output = ''.join(rmp_cmd_chars)
        
        return output        
        
        
    
